### PR TITLE
Add UserDataSerialization

### PR DESCRIPTION
**Pull request by JangkyuSeo**
### Purpose of this PR
Added the UserDataSerializer.
Register UserDataSerializer that inherits **UserDataSerializerBase** in the same way as Batcher and Streaming.
If you register, you can select like below figure.
image

The component for the serializer is added as shown in the figure below at build time.
image

Below is implement the UserDataSerializerBase.
``` 
public class EmptyUserDataSerializer : UserDataSerializerBase
{
    protected override void SerializeUserData(GameObject gameObject, HLODUserData data)
    {
     }
    protected override void DeserializeUserData(GameObject gameObject, HLODUserData data)
    {
    }
}
```

Below is a simple implementing UserDataSerializer.
```
public class TestUserDataSerializer : Unity.HLODSystem.Serializer.UserDataSerializerBase
{
#if UNITY_EDITOR
    [InitializeOnLoadMethod]
    static void Register()
    {
        Unity.HLODSystem.Serializer.UserDataSerializerTypes.RegisterType(typeof(TestUserDataSerializer));
    }
#endif

    protected override void SerializeUserData(GameObject gameObject, HLODUserData data)
    {
        var test = gameObject.GetComponent<TestSerialize>();
        if (test == null)
            return;

        data.IntDatas.AddData("Test", test.Value);
    }

    protected override void DeserializeUserData(GameObject gameObject, HLODUserData data)
    {
        if (data.IntDatas.HasData("Test") == false)
            return;

        var test = gameObject.AddComponent<TestSerialize>();
        test.Value = data.IntDatas.GetData("Test");
    }
}
```
---
### Release Notes
Add the UserDataSerialization

---
### Testing status
ManualTest on the AddressableSample.

---
### Overall Product Risks
**Technical Risk**: Low

**Halo Effect**: Medium

---

### CODE REVIEW

File: com.unity.hlod/Runtime/HLODUserData.cs [hunk 1]
```diff
@@ -0,0 +1,106 @@
+using System;
+using System.Collections.Generic;
+using Codice.Client.BaseCommands;
```
Comment (katana7755) [inline com.unity.hlod/Runtime/HLODUserData.cs:3]: I'm not sure what this namespace is for, but this makes building a player or assetbunldes fail.
When I removed the line, I couldn't see any compile error messages or build failure messages.

File: com.unity.hlod/Runtime/Serializer/UserDataSerializerBase.cs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        [HideInInspector]
+        private List<int> m_idList = new List<int>();
+        [SerializeField]
+        [HideInInspector]
+        private List<HLODUserData> m_userDataList = new List<HLODUserData>();
+        
+        public void SerializeUserData(int id, GameObject gameObject)
+        {
+            HLODUserData userData = new HLODUserData();
+            SerializeUserData(gameObject, userData);
```
Comment (katana7755) [inline com.unity.hlod/Runtime/Serializer/UserDataSerializerBase.cs:22]: This is more like a suggestion which is adding a branch to filter out and not insert the resulting user data into containers when the user data isn't written at all and stays at empty state.
By doing so, we can expect a certain amount of memory to be back!

File: com.unity.hlod/Runtime/HLODUserData.cs (hunk 1 above)
Comment (JangkyuSeo) [reply com.unity.hlod/Runtime/HLODUserData.cs]: Thanks!. I don't sure why it was there :) 
Fixied.

File: com.unity.hlod/Runtime/Serializer/UserDataSerializerBase.cs (hunk 2 above)
Comment (JangkyuSeo) [reply com.unity.hlod/Runtime/Serializer/UserDataSerializerBase.cs]: Thanks. Fixed. 

Also it had problem that it was not accumulate data.

**Review by katana7755 — APPROVED** [review]
Thanks for considering my suggestions seriously and also applying those to the system!!!
Now, all look great to me!!!

**Pull request merged**